spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpOutboundGateway.java

```diff
@@ -440,13 +454,25 @@ private final class AsyncReply {
 			this.connection = connection;
 			this.haveSemaphore = haveSemaphore;
 			if (async && remoteTimeout > 0) {
-				getTaskScheduler()
-						.schedule(() -> {
-							TcpOutboundGateway.this.pendingReplies.remove(connection.getConnectionId());
-							this.future.completeExceptionally(
-									new MessageTimeoutException(requestMessage, "Timed out waiting for response"));
-						}, Instant.now().plusMillis(remoteTimeout));
+				this.noResponseFuture =
+						getTaskScheduler()
+								.schedule(() -> {
+									cleanUp(this.haveSemaphore, this.connection, this.connection.getConnectionId());
```

Comment:
I think there's a new race here; we could end up releasing the semaphore twice because `onMessage` calls cleanup before completing the future.

Perhaps check if `pendingReplies` was actually removed in cleanup before releasing the semaphore?

Reply:
See some fixed I have pushed.